File: android-localwave/app/src/main/java/com/localwave/core/bluetooth/BlePacketFramer.kt
package com.localwave.core.bluetooth

import com.localwave.core.model.ConversationId
import com.localwave.core.model.PacketId
import com.localwave.core.model.TransportPacketKind
import com.localwave.core.protocol.ProtocolConstants
import java.nio.ByteBuffer
import java.util.UUID
import java.util.zip.CRC32

data class TransportPacket(
    val id: PacketId,
    val conversationId: ConversationId,
    val kind: TransportPacketKind,
    val chunkIndex: Int,
    val chunkCount: Int,
    val bodyLength: Long,
    val checksum: UInt,
    val body: ByteArray
) {
    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other !is TransportPacket) return false
        return id == other.id &&
            conversationId == other.conversationId &&
            kind == other.kind &&
            chunkIndex == other.chunkIndex &&
            chunkCount == other.chunkCount &&
            bodyLength == other.bodyLength &&
            checksum == other.checksum &&
            body.contentEquals(other.body)
    }

    override fun hashCode(): Int = id.hashCode()
}

data class ReassembledPacket(
    val packetId: PacketId,
    val conversationId: ConversationId,
    val kind: TransportPacketKind,
    val body: ByteArray
)

class BlePacketFramer(
    private val maxFrameBytes: Int = ProtocolConstants.DEFAULT_MAX_FRAME_BYTES,
    private val maxPayloadBytes: Int = ProtocolConstants.DEFAULT_MAX_PAYLOAD_BYTES
) {
    fun frame(
        body: ByteArray,
        kind: TransportPacketKind,
        packetId: UUID = UUID.randomUUID(),
        conversationId: UUID = UUID.randomUUID()
    ): List<TransportPacket> {
        require(body.size <= maxPayloadBytes) { "Messages are limited to $maxPayloadBytes bytes." }
        val maxChunkBytes = maxFrameBytes - ProtocolConstants.PACKET_HEADER_BYTES
        require(maxChunkBytes > 0) { "Max frame bytes must exceed header bytes." }
        val chunkCount = maxOf(1, (body.size + maxChunkBytes - 1) / maxChunkBytes)
        require(chunkCount <= UShort.MAX_VALUE.toInt()) { "Too many BLE chunks." }
        val checksum = crc32(body)
        return (0 until chunkCount).map { index ->
            val start = index * maxChunkBytes
            val end = minOf(start + maxChunkBytes, body.size)
            TransportPacket(
                id = packetId,
                conversationId = conversationId,
                kind = kind,
                chunkIndex = index,
                chunkCount = chunkCount,
                bodyLength = body.size.toLong(),
                checksum = checksum,
                body = if (start < end) body.copyOfRange(start, end) else ByteArray(0)
            )
        }
    }

    fun encode(packet: TransportPacket): ByteArray {
        val buffer = ByteBuffer.allocate(ProtocolConstants.PACKET_HEADER_BYTES + packet.body.size)
        buffer.put(ProtocolConstants.PACKET_VERSION.toByte())
        buffer.putUuid(packet.id)
        buffer.putUuid(packet.conversationId)
        buffer.put(packet.kind.wireValue.toByte())
        buffer.putShort(packet.chunkIndex.toShort())
        buffer.putShort(packet.chunkCount.toShort())
        buffer.putInt(packet.bodyLength.toInt())
        buffer.putInt(packet.checksum.toInt())
        buffer.put(packet.body)
        return buffer.array()
    }

    fun decode(data: ByteArray): TransportPacket {
        require(data.size >= ProtocolConstants.PACKET_HEADER_BYTES) { "Malformed packet." }
        val buffer = ByteBuffer.wrap(data)
        val version = buffer.get().toUByte()
        require(version == ProtocolConstants.PACKET_VERSION) { "Unsupported packet version." }
        val packetId = buffer.readUuid()
        val conversationId = buffer.readUuid()
        val kind = TransportPacketKind.fromWire(buffer.get().toUByte())
        val chunkIndex = buffer.short.toInt() and 0xffff
        val chunkCount = buffer.short.toInt() and 0xffff
        val bodyLength = buffer.int.toLong() and 0xffff_ffffL
        val checksum = buffer.int.toUInt()
        val body = ByteArray(buffer.remaining())
        buffer.get(body)
        require(chunkCount > 0 && chunkIndex < chunkCount) { "Malformed chunk indexes." }
        return TransportPacket(packetId, conversationId, kind, chunkIndex, chunkCount, bodyLength, checksum, body)
    }

    companion object {
        fun crc32(body: ByteArray): UInt {
            val crc = CRC32()
            crc.update(body)
            return crc.value.toUInt()
        }
    }
}

class PacketReassembler {
    private data class Accumulator(
        val conversationId: UUID,
        val kind: TransportPacketKind,
        val chunkCount: Int,
        val bodyLength: Long,
        val checksum: UInt,
        val chunks: MutableMap<Int, ByteArray> = mutableMapOf()
    )

    private val accumulators = mutableMapOf<UUID, Accumulator>()

    fun receive(packet: TransportPacket): ReassembledPacket? {
        require(packet.chunkCount > 0 && packet.chunkIndex < packet.chunkCount) { "Malformed packet." }
        val accumulator = accumulators.getOrPut(packet.id) {
            Accumulator(packet.conversationId, packet.kind, packet.chunkCount, packet.bodyLength, packet.checksum)
        }
        require(accumulator.conversationId == packet.conversationId)
        require(accumulator.kind == packet.kind)
        require(accumulator.chunkCount == packet.chunkCount)
        require(accumulator.bodyLength == packet.bodyLength)
        require(accumulator.checksum == packet.checksum)
        accumulator.chunks[packet.chunkIndex] = packet.body
        if (accumulator.chunks.size != accumulator.chunkCount) return null
        val body = ByteArray(accumulator.bodyLength.toInt())
        var offset = 0
        for (index in 0 until accumulator.chunkCount) {
            val chunk = accumulator.chunks[index] ?: return null
            chunk.copyInto(body, offset)
            offset += chunk.size
        }
        accumulators.remove(packet.id)
        require(offset == body.size && BlePacketFramer.crc32(body) == accumulator.checksum) { "Malformed packet." }
        return ReassembledPacket(packet.id, accumulator.conversationId, accumulator.kind, body)
    }
}

private fun ByteBuffer.putUuid(uuid: UUID) {
    putLong(uuid.mostSignificantBits)
    putLong(uuid.leastSignificantBits)
}

private fun ByteBuffer.readUuid(): UUID = UUID(long, long)

fun ByteArray.toHex(): String = joinToString("") { "%02x".format(it.toInt() and 0xff) }

File: ios-localwave/LocalWave/Core/Bluetooth/BLEPacketFramer.swift
import Foundation

public struct ReassembledPacket: Equatable, Sendable {
    public var packetId: PacketID
    public var conversationId: ConversationID
    public var kind: TransportPacketKind
    public var body: Data
}

public struct BLEPacketFramer: Sendable {
    public static let headerBytes = 46
    public static let defaultMaxFrameBytes = 182
    public static let defaultMaxPayloadBytes = 16 * 1024

    public var maxFrameBytes: Int
    public var maxPayloadBytes: Int

    public init(maxFrameBytes: Int = Self.defaultMaxFrameBytes, maxPayloadBytes: Int = Self.defaultMaxPayloadBytes) {
        self.maxFrameBytes = maxFrameBytes
        self.maxPayloadBytes = maxPayloadBytes
    }

    public func frame(
        body: Data,
        kind: TransportPacketKind,
        packetId: PacketID = UUID(),
        conversationId: ConversationID
    ) throws -> [TransportPacket] {
        guard body.count <= maxPayloadBytes else {
            throw LocalWaveError.oversizedMessage(limitBytes: maxPayloadBytes)
        }
        let maxChunkBytes = maxFrameBytes - Self.headerBytes
        guard maxChunkBytes > 0 else { throw LocalWaveError.malformedPacket }

        let chunkCount = max(1, Int(ceil(Double(body.count) / Double(maxChunkBytes))))
        guard chunkCount <= Int(UInt16.max), body.count <= Int(UInt32.max) else {
            throw LocalWaveError.oversizedMessage(limitBytes: maxPayloadBytes)
        }

        let checksum = CRC32.checksum(body)
        return (0..<chunkCount).map { index in
            let start = index * maxChunkBytes
            let end = min(start + maxChunkBytes, body.count)
            let chunkBody = start < end ? body.subdata(in: start..<end) : Data()
            return TransportPacket(
                id: packetId,
                conversationId: conversationId,
                kind: kind,
                chunkIndex: UInt16(index),
                chunkCount: UInt16(chunkCount),
                bodyLength: UInt32(body.count),
                checksum: checksum,
                body: chunkBody
            )
        }
    }

    public func chunk(
        body: Data,
        kind: TransportPacketKind,
        packetId: PacketID = UUID(),
        conversationId: ConversationID = UUID()
    ) throws -> [TransportPacket] {
        try frame(body: body, kind: kind, packetId: packetId, conversationId: conversationId)
    }

    public func reassemble(_ packets: [TransportPacket]) throws -> Data {
        let reassembler = BLEPacketReassembler()
        for packet in packets.sorted(by: { $0.chunkIndex < $1.chunkIndex }) {
            if let result = try reassembler.receive(packet) {
                return result.body
            }
        }
        throw LocalWaveError.malformedPacket
    }

    public func encode(_ packet: TransportPacket) -> Data {
        var data = Data()
        data.append(TransportPacket.version)
        data.append(packet.id.uuidBytes)
        data.append(packet.conversationId.uuidBytes)
        data.append(packet.kind.rawValue)
        data.append(packet.chunkIndex.bigEndianData)
        data.append(packet.chunkCount.bigEndianData)
        data.append(packet.bodyLength.bigEndianData)
        data.append(packet.checksum.bigEndianData)
        data.append(packet.body)
        return data
    }

    public func decode(_ data: Data) throws -> TransportPacket {
        var cursor = DataCursor(data: data)
        let version = try cursor.readUInt8()
        guard version == TransportPacket.version else { throw LocalWaveError.malformedPacket }
        let packetId = try cursor.readUUID()
        let conversationId = try cursor.readUUID()
        guard let kind = TransportPacketKind(rawValue: try cursor.readUInt8()) else {
            throw LocalWaveError.malformedPacket
        }
        let chunkIndex = try cursor.readUInt16()
        let chunkCount = try cursor.readUInt16()
        let bodyLength = try cursor.readUInt32()
        let checksum = try cursor.readUInt32()
        let body = cursor.remainingData()

        guard chunkCount > 0, chunkIndex < chunkCount else {
            throw LocalWaveError.malformedPacket
        }

        return TransportPacket(
            id: packetId,
            conversationId: conversationId,
            kind: kind,
            chunkIndex: chunkIndex,
            chunkCount: chunkCount,
            bodyLength: bodyLength,
            checksum: checksum,
            body: body
        )
    }
}

public final class BLEPacketReassembler {
    private struct Accumulator {
        var conversationId: ConversationID
        var kind: TransportPacketKind
        var chunkCount: UInt16
        var bodyLength: UInt32
        var checksum: UInt32
        var chunks: [UInt16: Data]
    }

    private var accumulators: [PacketID: Accumulator] = [:]

    public init() {}

    public func receive(_ packet: TransportPacket) throws -> ReassembledPacket? {
        guard packet.chunkCount > 0, packet.chunkIndex < packet.chunkCount else {
            throw LocalWaveError.malformedPacket
        }

        var accumulator = accumulators[packet.id] ?? Accumulator(
            conversationId: packet.conversationId,
            kind: packet.kind,
            chunkCount: packet.chunkCount,
            bodyLength: packet.bodyLength,
            checksum: packet.checksum,
            chunks: [:]
        )

        guard accumulator.conversationId == packet.conversationId,
              accumulator.kind == packet.kind,
              accumulator.chunkCount == packet.chunkCount,
              accumulator.bodyLength == packet.bodyLength,
              accumulator.checksum == packet.checksum else {
            throw LocalWaveError.malformedPacket
        }

        accumulator.chunks[packet.chunkIndex] = packet.body
        accumulators[packet.id] = accumulator

        guard accumulator.chunks.count == Int(accumulator.chunkCount) else {
            return nil
        }

        var body = Data()
        for index in 0..<accumulator.chunkCount {
            guard let chunk = accumulator.chunks[index] else {
                return nil
            }
            body.append(chunk)
        }

        accumulators[packet.id] = nil
        guard body.count == Int(accumulator.bodyLength), CRC32.checksum(body) == accumulator.checksum else {
            throw LocalWaveError.malformedPacket
        }

        return ReassembledPacket(
            packetId: packet.id,
            conversationId: accumulator.conversationId,
            kind: accumulator.kind,
            body: body
        )
    }
}

enum CRC32 {
    private static let table: [UInt32] = (0..<256).map { index in
        var crc = UInt32(index)
        for _ in 0..<8 {
            crc = (crc & 1) == 1 ? (0xEDB88320 ^ (crc >> 1)) : (crc >> 1)
        }
        return crc
    }

    static func checksum(_ data: Data) -> UInt32 {
        var crc: UInt32 = 0xFFFF_FFFF
        for byte in data {
            crc = table[Int((crc ^ UInt32(byte)) & 0xFF)] ^ (crc >> 8)
        }
        return crc ^ 0xFFFF_FFFF
    }
}

private struct DataCursor {
    private let data: Data
    private var offset = 0

    init(data: Data) {
        self.data = data
    }

    mutating func readUInt8() throws -> UInt8 {
        guard offset < data.count else { throw LocalWaveError.malformedPacket }
        defer { offset += 1 }
        return data[offset]
    }

    mutating func readUInt16() throws -> UInt16 {
        let bytes = Array(try readBytes(count: 2))
        return (UInt16(bytes[0]) << 8) | UInt16(bytes[1])
    }

    mutating func readUInt32() throws -> UInt32 {
        let bytes = Array(try readBytes(count: 4))
        return (UInt32(bytes[0]) << 24) | (UInt32(bytes[1]) << 16) | (UInt32(bytes[2]) << 8) | UInt32(bytes[3])
    }

    mutating func readUUID() throws -> UUID {
        let bytes = Array(try readBytes(count: 16))
        return bytes.withUnsafeBufferPointer { pointer in
            NSUUID(uuidBytes: pointer.baseAddress!) as UUID
        }
    }

    mutating func readBytes(count: Int) throws -> Data {
        guard offset + count <= data.count else { throw LocalWaveError.malformedPacket }
        defer { offset += count }
        return data.subdata(in: offset..<(offset + count))
    }

    func remainingData() -> Data {
        data.subdata(in: offset..<data.count)
    }
}

private extension UUID {
    var uuidBytes: Data {
        var bytes = uuid
        return withUnsafeBytes(of: &bytes) { Data($0) }
    }
}

private extension FixedWidthInteger {
    var bigEndianData: Data {
        var value = self.bigEndian
        return withUnsafeBytes(of: &value) { Data($0) }
    }
}
